### PR TITLE
Implement QoS: liveliness, deadline, lifespan

### PR DESCRIPTION
**Pull request by ross-desmond**
**Summary**
Provide implementation of QoS changes for rmw events. If applied this PR will
* ADD support for QoS Liveliness Lost, Offered Deadline Missed, Liveliness Changed, and Requested Deadline Missed
* ADD manual Liveliness assertion
* ADD conversion methods to map from Opensplice QoS DDS events to RMW QoS events 

Connects to ros2/rmw#171

### CODE REVIEW

**Comment by wjwwood** [comment]
I'm going to hold off reviewing this one until some of the common nitpick's I pointed out in https://github.com/ros2/rmw_connext/pull/352#pullrequestreview-229943218 can be checked for here too.

**Comment by wjwwood** [comment]
It doesn't look like any commits have been pushed here since my last comment. I think most of the things I mentioned on the `rmw_connext` issue have been addressed there, but not here, e.g.:

- wrong doxygen style: https://github.com/ros2/rmw_opensplice/pull/266/files#diff-e8093a1e88afed30e3ccf68994e4dff4R21
- weird indentation: https://github.com/ros2/rmw_opensplice/pull/266/files#diff-08789ebd264910243b78cdf82cbadc95R425

Maybe you guys could take a pass over this before I do a full review?

**Comment by mm318** [comment]
Hi @wjwwood, the common nitpicks should be corrected now. Thanks!

File: rmw_opensplice_cpp/src/rmw_wait.cpp [hunk 1]
```diff
@@ -196,6 +272,23 @@ rmw_wait(
...
 
+  std::unordered_set<DDS::StatusCondition *> status_conditions;
+  // gather all status conditions with set masks
+  rmw_ret_t ret_code = __gather_event_conditions(events, status_conditions);
+  if (ret_code != RMW_RET_OK) {
+    return ret_code;
+  }
+  if (!status_conditions.empty()) {
+    // enable a status condition for each event
+    for (auto * status_condition : status_conditions) {
```
Comment (wjwwood) [inline rmw_opensplice_cpp/src/rmw_wait.cpp]: I mentioned this elsewhere (different pr), but the `*` isn't strictly needed. If you want to leave it for clarity, that's fine, but I just wanted to point it out.

File: rmw_opensplice_cpp/src/types.cpp [hunk 2]
```diff
@@ -342,3 +345,105 @@ size_t OpenSpliceSubscriberListener::current_count() const
...
+          static_cast<rmw_liveliness_lost_status_t *>(event);
+        rmw_liveliness_lost->total_count = liveliness_lost.total_count;
+        rmw_liveliness_lost->total_count_change = liveliness_lost.total_count_change;
+        break;
+      }
+    case DDS::OFFERED_DEADLINE_MISSED_STATUS:
+      {
+        DDS::OfferedDeadlineMissedStatus offered_deadline_missed;
+        DDS::ReturnCode_t dds_return_code = topic_writer
+          ->get_offered_deadline_missed_status(offered_deadline_missed);
```
Comment (wjwwood) [inline rmw_opensplice_cpp/src/types.cpp]: nitpick: weird place to wrap, might want to look for them, `     ->` might be something to search for.

File: rmw_opensplice_cpp/src/types.cpp [hunk 3]
```diff
@@ -342,3 +345,105 @@ size_t OpenSpliceSubscriberListener::current_count() const
...
+      {
+        DDS::LivelinessLostStatus liveliness_lost;
+        DDS::ReturnCode_t dds_return_code =
+          topic_writer->get_liveliness_lost_status(liveliness_lost);
+        rmw_ret_t from_dds = check_dds_ret_code(dds_return_code);
+        if (from_dds != RMW_RET_OK) {
+          return from_dds;
+        }
+
+        rmw_liveliness_lost_status_t * rmw_liveliness_lost =
```
Comment (wjwwood) [inline rmw_opensplice_cpp/src/types.cpp]: nitpick: good place to use `auto`.

File: rmw_opensplice_cpp/src/event_converter.hpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * \param event_t input rmw event to check
+ * \return true if there is an RMW to DDS_StatusKind mapping, false otherwise
+ */
+bool is_event_supported(const rmw_event_type_t event_t);
+
+/// Assign the input DDS return code to its corresponding RMW return code.
+/**
+  * \param dds_return_code input DDS return code
+  * \return to_return the corresponding rmw_ret_t that maps to the input DDS_ReturnCode_t. By
+  *         default RMW_RET_ERROR is returned if no corresponding rmw_ret_t is not defined.
```
[the remark this answers is not in the document]
Comment (mm318) [reply rmw_opensplice_cpp/src/event_converter.hpp]: Done.

File: rmw_opensplice_cpp/src/rmw_wait.cpp (hunk 1 above)
Comment (mm318) [reply rmw_opensplice_cpp/src/rmw_wait.cpp]: I have removed the `*` to be consistent with `rmw_connext_shared_cpp`.

File: rmw_opensplice_cpp/src/types.cpp (hunk 2 above)
Comment (mm318) [reply rmw_opensplice_cpp/src/types.cpp]: Fixed.

I did a search for ` ->` and it looks like that was the last one.

File: rmw_opensplice_cpp/src/types.cpp (hunk 3 above)
Comment (mm318) [reply rmw_opensplice_cpp/src/types.cpp]: Done.

**Comment by dirk-thomas** [comment]
This PR introduces a few compiler warnings, see https://ci.ros2.org/job/ci_packaging_windows/105/warnings43Result/ Please address them in another PR rather sooner than later.

**Comment by mm318** [comment]
Opened another PR for addressing the new compiler warnings at https://github.com/ros2/rmw_opensplice/pull/269.

**Pull request merged**